sql: add deprecation path for pg incompatible database privileges

RichardJCai opened the pull request:
Release note (sql change): Granting SELECT, UPDATE, INSERT, DELETE
on databases is being deprecated. The syntax is still supported however
it is automatically converted to the equivalent ALTER DEFAULT PRIVILEGES
FOR ALL ROLES command. The user is given a notice that the privilege is
incompatible and automatically being converted to an ALTER DEFAULT PRIVILEGE
FOR ALL ROLES command.

Example:
> GRANT SELECT, UPDATE ON DATABASE defaultdb TO demo;
NOTICE: granting SELECT, UPDATE is going to be deprecated in 22.1, please use ALTER DEFAULT PRIVILEGES FOR ALL ROLES instead in the futurethe incompatible privileges were not granted, "ALTER DEFAULT PRIVILEGES FOR ALL ROLES GRANT SELECT, UPDATE ON TABLES TO demo" was automatically executed instead
GRANT

Resolves https://github.com/cockroachdb/cockroach/issues/67378

Refer only to last commit.
sql: add deprecation path for pg incompatible database privileges

RichardJCai committed:
sql: add deprecation path for pg incompatible database privileges

Release note (sql change): Granting SELECT, UPDATE, INSERT, DELETE
on databases is being deprecated. The syntax is still supported however
it is automatically converted to the equivalent ALTER DEFAULT PRIVILEGES
FOR ALL ROLES command. The user is given a notice that the privilege is
incompatible and automatically being converted to an ALTER DEFAULT PRIVILEGE
FOR ALL ROLES command.

Example:
> GRANT SELECT, UPDATE ON DATABASE defaultdb TO demo;
NOTICE: granting SELECT, UPDATE is going to be deprecated in 22.1, please use ALTER DEFAULT PRIVILEGES FOR ALL ROLES instead in the futurethe incompatible privileges were not granted, "ALTER DEFAULT PRIVILEGES FOR ALL ROLES GRANT SELECT, UPDATE ON TABLES TO demo" was automatically executed instead
GRANT

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68391)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, and @RichardJCai)

___
*[pkg/sql/alter_default_privileges.go, line 171 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mglrtyc6a31Yo8A-PJB:-Mglrtyc6a31Yo8A-PJC:bbfsh9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/alter_default_privileges.go#L171)):*
> ```Go
> 			eventDetails.RevokedPrivileges = privileges.SortedNames()
> 		}
> 		event := eventpb.AlterDefaultPrivileges{
> ```

i think the event log change is def important, but was it supposed to be part of this PR?

(not a big deal, just want to confirm that the code doesn't get misplaced)
___
*[pkg/sql/grant_revoke.go, line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MglsS1sEbA_oMxpVvk9:-MglsS1sEbA_oMxpVvkA:b-313nbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L51)):*
> ```Go
> 		isGrant:      true,
> 		targets:      n.Targets,
> 		grantees:     n.Grantees,
> ```

why use the non-normalized names here? why was the normalization moved to startExec?
___
*[pkg/sql/grant_revoke.go, line 149 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mglzs180vJlEQeqqaK5:-Mglzs180vJlEQeqqaK6:b-ex9ywi) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L149)):*
> ```Go
> 			// The objectType is always Tables since the incompatible pg privileges
> 			// are (SELECT, INSERT, UPDATE, DELETE) which were previously allowed
> 			// for the sake of inheriting privileges from the table to the db.
> ```

nit: from the DB to the table?
___
*[pkg/sql/grant_revoke.go, line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm-32K6HKE9nyrQrJ1:-Mgm-32K6HKE9nyrQrJ2:b-xmlf2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L158)):*
> ```Go
> 				// TODO(richardjcai): Not sure if this needs to be redacted.
> 				translatedStatement = fmt.Sprintf("ALTER DEFAULT PRIVILEGES "+
> 					"FOR ALL ROLES GRANT %s ON TABLES TO %s; in database %s",
> ```

nit: there's a semicolon after the second `%s`

also make `in database` caps?
___
*[pkg/sql/grant_revoke.go, line 168 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm-ADVDHb66DC3UqX-:-Mgm-ADW2t7Qaw8kySn1:b-9uvqrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L168)):*
> ```Go
> 				}, incompatiblePrivs, n.grantees, objectType)
> 				translatedStatement = fmt.Sprintf("ALTER DEFAULT PRIVILEGES "+
> 					"FOR ALL ROLES REVOKE %s ON TABLES FROM %s; in database %s",
> ```

nit: semicolon and caps
___
*[pkg/sql/grant_revoke.go, line 175 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm-E1YExjUTPueuvsy:-Mgm-E1YExjUTPueuvsz:bsvikvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L175)):*
> ```Go
> 			if len(incompatiblePrivs) > 0 {
> 				p.BufferClientNotice(ctx, pgnotice.Newf(
> 					"granting %s is going to be deprecated in 22.1, please use ALTER DEFAULT PRIVILEGES FOR ALL ROLES instead in the future"+
> ```

i think the message can say `granting %s is deprecated, please use ALTER DEFAULT PRIVILEGES FOR ALL ROLES.\n` (note the newline, and avoid mentioning the exact version number)

and instead of "the incompatible privileges were not granted" it could again use the `incompatiblePrivs` list

and maybe say "the statement was automatically translated to %s"
___
*[pkg/sql/privilege/privilege.go, line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm0F246xBkA-q_b06W:-Mgm0F246xBkA-q_b06X:bo9i0qq) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/privilege/privilege.go#L82)):*
> ```Go
> // deprecated and instead run the equivalent ALTER DEFAULT PRIVILEGES FOR ALL
> // ROLES.
> // TODO(richardcai): Remove this and the syntax for granting these privileges
> ```

can you file a github issue for this and refer to it here?


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, and @rafiss)

___
*[pkg/sql/alter_default_privileges.go, line 171 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mglrtyc6a31Yo8A-PJB:-Mgqps2D40s6whw9aBbd:b-sqe9nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/alter_default_privileges.go#L171)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think the event log change is def important, but was it supposed to be part of this PR?

(not a big deal, just want to confirm that the code doesn't get misplaced)
</blockquote></details>

Yeah I intended to add it here
___
*[pkg/sql/grant_revoke.go, line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MglsS1sEbA_oMxpVvk9:-Mgqq2xyDeQpzVg5ngY9:b-ntrgq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L51)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why use the non-normalized names here? why was the normalization moved to startExec?
</blockquote></details>

Only cause `Grant and RevokeDefaultPrivileges` took the non-normalized name, in hindsight I can just change them to take the normalized name, I'll do that instead
___
*[pkg/sql/grant_revoke.go, line 149 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mglzs180vJlEQeqqaK5:-MgqrGA939DrqXXgYm3X:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L149)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: from the DB to the table?
</blockquote></details>

done
___
*[pkg/sql/grant_revoke.go, line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm-32K6HKE9nyrQrJ1:-Mgqr_w7-ftf7u3h2B5M:b86tojv) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L158)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: there's a semicolon after the second `%s`

also make `in database` caps?
</blockquote></details>

I added brackets around IN DATABASE, to make it clear it isn't part of the statement but rather which database the statement was executed in.

There is no `ALTER DEFAULT PRIVILEGES IN DATABASE ...` so the user has to use the database
___
*[pkg/sql/grant_revoke.go, line 175 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm-E1YExjUTPueuvsy:-MgqsIX5DW906XEjI_lw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L175)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think the message can say `granting %s is deprecated, please use ALTER DEFAULT PRIVILEGES FOR ALL ROLES.\n` (note the newline, and avoid mentioning the exact version number)

and instead of "the incompatible privileges were not granted" it could again use the `incompatiblePrivs` list

and maybe say "the statement was automatically translated to %s"
</blockquote></details>

Done.
___
*[pkg/sql/privilege/privilege.go, line 82 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm0F246xBkA-q_b06W:-MgqsnHs0Wu5auyDTU0R:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/privilege/privilege.go#L82)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you file a github issue for this and refer to it here?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
looks good to me! feel free to merge when you want to

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-:-Mgt0YZ10qAErMyzcBj6:b6vtx59)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/grant_revoke.go, line 158 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-Mgm-32K6HKE9nyrQrJ1:-Mgt-jff3gbaMxJUWCZe:b7gulor) ([raw file](https://github.com/cockroachdb/cockroach/blob/33f6684cc5c6300c022cc67c8020c875fc57c67d/pkg/sql/grant_revoke.go#L158)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I added brackets around IN DATABASE, to make it clear it isn't part of the statement but rather which database the statement was executed in.

There is no `ALTER DEFAULT PRIVILEGES IN DATABASE ...` so the user has to use the database
</blockquote></details>

ah i see; i misunderstood then. maybe it's better as `SET DATABASE = %q; ALTER DEFAULT PRIVILEGES FOR ALL ROLES GRANT ...`

up to you though whichever you feel is most clear


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
LGTM modulo doc fix

> Reviewed 7 of 10 files at r5, 23 of 23 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-:-MguPKc078t8xdqAxR0v:b-80colc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @rafiss, and @RichardJCai)

___
*[pkg/util/log/eventpb/privilege_events.proto, line 133 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MguP78S79J8K5ZkfxLX:-MguP78S79J8K5ZkfxLY:b-8rz4uh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bfa0b557638a53ca55c344335c1b6ac0e8f10ab/pkg/util/log/eventpb/privilege_events.proto#L133)):*
> ```protobuf
>   // The name of the affected database.
>   string database_name = 4 [(gogoproto.jsontag) = ",omitempty"];
>   // The role having it's default privileges altered.
> ```

"its"
___
*[pkg/util/log/eventpb/privilege_events.proto, line 135 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MguP7vZ9Hvl95bB63S1:-MguP7vZ9Hvl95bB63S2:b-4owoy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bfa0b557638a53ca55c344335c1b6ac0e8f10ab/pkg/util/log/eventpb/privilege_events.proto#L135)):*
> ```protobuf
>   // The role having it's default privileges altered.
>   string role_name = 5 [(gogoproto.jsontag) = ",omitempty"];
>   bool for_all_roles = 6 [(gogoproto.jsontag) = ",omitempty"];;
> ```

Explain the field.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, and @rafiss)

___
*[pkg/util/log/eventpb/privilege_events.proto, line 133 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MguP78S79J8K5ZkfxLX:-Mh0RETT75u5hxMf1mKu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bfa0b557638a53ca55c344335c1b6ac0e8f10ab/pkg/util/log/eventpb/privilege_events.proto#L133)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"its"
</blockquote></details>

Done.
___
*[pkg/util/log/eventpb/privilege_events.proto, line 135 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MguP7vZ9Hvl95bB63S1:-Mh0REkdESD92n73iPT2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bfa0b557638a53ca55c344335c1b6ac0e8f10ab/pkg/util/log/eventpb/privilege_events.proto#L135)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Explain the field.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 2 of 6 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/grant_revoke.go, line 107 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhFOKBODZyCHxolE9sc:-MhFOKBODZyCHxolE9sd:b-tbukzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/78960bb937b56ce4c23fe28c7257beb00712a62d/pkg/sql/grant_revoke.go#L107)):*
> ```Go
> func (n *changePrivilegesNode) ReadingOwnWrites() {}
> 
> func (n *changePrivilegesNode) startExec(params runParams) error {
> ```

from a high level, why are we doing this here and not in `Revoke` and `Grant`. It feels like the right way to frame this is converting the ast node to a different type of plan node as opposed to changing the old plan node to support this behavior. Could we just construct an ast node in place here for the default privilege change and call that function? 
___
*[pkg/sql/grant_revoke.go, line 133 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhFFWWy0qj61zftezyx:-MhFFWWzB8GHeqXSOZBk:bxkybav) ([raw file](https://github.com/cockroachdb/cockroach/blob/78960bb937b56ce4c23fe28c7257beb00712a62d/pkg/sql/grant_revoke.go#L133)):*
> ```
> 	// Convert PGIncompatibleDatabase privileges to default privileges instead.
> 	if len(n.targets.Databases) != 0 {
> ```
thoughts on pulling this into a helper?


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/grant_revoke.go, line 107 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhFOKBODZyCHxolE9sc:-MhJxTDL2sQGvJLZsrAo:b-nm8gxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/78960bb937b56ce4c23fe28c7257beb00712a62d/pkg/sql/grant_revoke.go#L107)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

from a high level, why are we doing this here and not in `Revoke` and `Grant`. It feels like the right way to frame this is converting the ast node to a different type of plan node as opposed to changing the old plan node to support this behavior. Could we just construct an ast node in place here for the default privilege change and call that function? 
</blockquote></details>

Constructing an AST node and calling the plan node makes sense. I guess one thing I'm unsure about if we put it in `Revoke` and `Grant` is do we directly call `startExec` on the new plan node, have it execute, then `Revoke` and `Grant` return it's own planNode for `changePrivilegesNode` and have the execute after?


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/grant_revoke.go, line 107 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhFOKBODZyCHxolE9sc:-MhK8OfAAsyF_3XW9GVh:ba60ine) ([raw file](https://github.com/cockroachdb/cockroach/blob/78960bb937b56ce4c23fe28c7257beb00712a62d/pkg/sql/grant_revoke.go#L107)):*
>It feels like the right way to frame this is converting the ast node to a different type of plan node as opposed to changing the old plan node to support this behavior. 

Ah one thing to note is that we sometimes have to run both plan nodes, ie `GRANT SELECT, CONNECT ON DATABASE db TO role1` we have to perform both a `GRANT` (for CONNECT which is valid) and `ALTER DEFAULT PRIVILEGES` (for SELECT which is invalid), so it's not just a matter of deciding which plan node to return.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, and @rafiss)

___
*[pkg/sql/grant_revoke.go, line 133 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhFFWWy0qj61zftezyx:-MhKNMoH7-WfKk2erhnZ:b-dcl5ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/78960bb937b56ce4c23fe28c7257beb00712a62d/pkg/sql/grant_revoke.go#L133)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Convert PGIncompatibleDatabase privileges to default privileges instead.
> 	if len(n.targets.Databases) != 0 {
> ```
thoughts on pulling this into a helper?
</blockquote></details>

Pulled this into helper


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Seems like there's a bit of testing to fix up.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-:-MhKupAGBzdoPlSikv-q:b8o8ibh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/grant_revoke.go, line 107 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhFOKBODZyCHxolE9sc:-MhKu9Ks8OlxTfqvBsb2:bsdl082) ([raw file](https://github.com/cockroachdb/cockroach/blob/78960bb937b56ce4c23fe28c7257beb00712a62d/pkg/sql/grant_revoke.go#L107)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

>It feels like the right way to frame this is converting the ast node to a different type of plan node as opposed to changing the old plan node to support this behavior. 

Ah one thing to note is that we sometimes have to run both plan nodes, ie `GRANT SELECT, CONNECT ON DATABASE db TO role1` we have to perform both a `GRANT` (for CONNECT which is valid) and `ALTER DEFAULT PRIVILEGES` (for SELECT which is invalid), so it's not just a matter of deciding which plan node to return.
</blockquote></details>

Aha, interesting, makes sense. 
___
*[pkg/sql/grant_revoke.go, line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhKuH_t5-AnrVb2f030:-MhKuH_t5-AnrVb2f031:bpbzgkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdddbb0be3cdd7c7ff316119ff9d593dcb3fffa4/pkg/sql/grant_revoke.go#L133)):*
> ```
> 	if n.isGrant {
> 		if err != nil {
> 			return err
> ```
what is this about?
___
*[pkg/sql/grant_revoke.go, line 424 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhKudAYEUdUwJLEiXe5:-MhKudAYEUdUwJLEiXe6:b-qhjgrn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdddbb0be3cdd7c7ff316119ff9d593dcb3fffa4/pkg/sql/grant_revoke.go#L424)):*
<details><summary><i alt="summary">Quoted 47 lines of code…</i></summary>

> ```
> 			if n.isGrant {
> 				alterDefaultPrivilegesASTNode := tree.AlterDefaultPrivileges{
> 					ForAllRoles: true,
> 					IsGrant:     n.isGrant,
> 					Database:    &database,
> 					Grant: tree.AbbreviatedGrant{
> 						Grantees:   grantees,
> 						Privileges: incompatiblePrivs,
> 						Target:     objectType,
> 					},
> 				}
> 				alterDefaultPrivilegesNode, err := p.alterDefaultPrivileges(
> 					ctx, &alterDefaultPrivilegesASTNode,
> 				)
> 				if err != nil {
> 					return nil, err
> 				}
> 				translatedStatement = fmt.Sprintf("USE %s; %s;", database.Normalize(),
> 					alterDefaultPrivilegesASTNode.String())
> 				if err := alterDefaultPrivilegesNode.startExec(p.RunParams(ctx)); err != nil {
> 					return nil, err
> 				}
> 			} else if !n.isGrant {
> 				// If revoke is specified, we need to revoke both existing privileges
> 				// on the database and also default privileges.
> 				alterDefaultPrivilegesASTNode := tree.AlterDefaultPrivileges{
> 					ForAllRoles: true,
> 					IsGrant:     n.isGrant,
> 					Database:    &database,
> 					Revoke: tree.AbbreviatedRevoke{
> 						Grantees:   grantees,
> 						Privileges: incompatiblePrivs,
> 						Target:     objectType,
> 					},
> 				}
> 				alterDefaultPrivilegesNode, err := p.alterDefaultPrivileges(
> 					ctx, &alterDefaultPrivilegesASTNode,
> 				)
> 				if err != nil {
> 					return nil, err
> 				}
> 				translatedStatement = fmt.Sprintf("USE %s; %s;", database.Normalize(),
> 					alterDefaultPrivilegesASTNode.String())
> 				if err := alterDefaultPrivilegesNode.startExec(p.RunParams(ctx)); err != nil {
> 					return nil, err
> 				}
> 			}
> ```
</details>

this code could be deduplicated a bit


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, @rafiss, and @RichardJCai)

___
*[pkg/sql/grant_revoke.go, line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhKuH_t5-AnrVb2f030:-MhL6jYrEy6Me_jsp0aX:b-z2j0kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdddbb0be3cdd7c7ff316119ff9d593dcb3fffa4/pkg/sql/grant_revoke.go#L133)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	if n.isGrant {
> 		if err != nil {
> 			return err
> ```
what is this about?
</blockquote></details>

Removed, didn't fully delete block
___
*[pkg/sql/grant_revoke.go, line 424 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhKudAYEUdUwJLEiXe5:-MhL6o52AlmHRaoDpFQR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdddbb0be3cdd7c7ff316119ff9d593dcb3fffa4/pkg/sql/grant_revoke.go#L424)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 			if n.isGrant {
> 				alterDefaultPrivilegesASTNode := tree.AlterDefaultPrivileges{
> 					ForAllRoles: true,
> 					IsGrant:     n.isGrant,
> 					Database:    &database,
> 					Grant: tree.AbbreviatedGrant{
> 						Grantees:   grantees,
> 						Privileges: incompatiblePrivs,
> 						Target:     objectType,
> 					},
> 				}
> 				alterDefaultPrivilegesNode, err := p.alterDefaultPrivileges(
> 					ctx, &alterDefaultPrivilegesASTNode,
> 				)
> 				if err != nil {
> 					return nil, err
> 				}
> 				translatedStatement = fmt.Sprintf("USE %s; %s;", database.Normalize(),
> 					alterDefaultPrivilegesASTNode.String())
> 				if err := alterDefaultPrivilegesNode.startExec(p.RunParams(ctx)); err != nil {
> 					return nil, err
> 				}
> 			} else if !n.isGrant {
> 				// If revoke is specified, we need to revoke both existing privileges
> 				// on the database and also default privileges.
> 				alterDefaultPrivilegesASTNode := tree.AlterDefaultPrivileges{
> 					ForAllRoles: true,
> 					IsGrant:     n.isGrant,
> 					Database:    &database,
> 					Revoke: tree.AbbreviatedRevoke{
> 						Grantees:   grantees,
> 						Privileges: incompatiblePrivs,
> 						Target:     objectType,
> 					},
> 				}
> 				alterDefaultPrivilegesNode, err := p.alterDefaultPrivileges(
> 					ctx, &alterDefaultPrivilegesASTNode,
> 				)
> 				if err != nil {
> 					return nil, err
> 				}
> 				translatedStatement = fmt.Sprintf("USE %s; %s;", database.Normalize(),
> 					alterDefaultPrivilegesASTNode.String())
> 				if err := alterDefaultPrivilegesNode.startExec(p.RunParams(ctx)); err != nil {
> 					return nil, err
> 				}
> 			}
> ```
this code could be deduplicated a bit
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RichardJCai committed:
backupccl: optionally check CONNECT privilege on database for backup

Release note (enterprise change): Check for CONNECT privilege on the database
to allow for backing up the database.

SELECT is still checked, however as SELECT is being deprecated as a database
privilege, CONNECT is now also checked. Existing users with SELECT on the
database can still backup the database, however moving forward, it is
recommended to GRANT CONNECT on the database.

RichardJCai commented:
For @cockroachdb/bulk-prs, the second commit I made a change to the privilege check for backing up databases.

As SELECT is being deprecated on databases, CONNECT privilege is required instead. (SELECT is not a privilege for databases in pg, CONNECT is).

dt commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -674,7 +674,16 @@ func checkPrivilegesForBackup(
+			// In the meanwhile, we still allow backup if the user has SELECT.
+			// TODO(richardjcai): Remove this check for SELECT in 22.1.
+			selectErr := p.CheckPrivilege(ctx, desc, privilege.SELECT)
+			connectErr := p.CheckPrivilege(ctx, desc, privilege.CONNECT)
```
why not move one of these -- maybe the SELECT one -- inside the err != nil check for the other to avoid the double check if the first succeeds?

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68391)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @knz, @rafiss, and @RichardJCai)

___
*[pkg/ccl/backupccl/backup_planning.go, line 682 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68391#-MhOzmOrM1kudFaFTZIT-r12-682:-MhPQ7EtDnjIAJUBa_KU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdca049d4567b509c433b9c83fb46521e11a365a/pkg/ccl/backupccl/backup_planning.go#L682)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

why not move one of these -- maybe the SELECT one -- inside the err != nil check for the other to avoid the double check if the first succeeds?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
backup change LGTM

RichardJCai commented:
Thanks for the reviews!

bors r+
